sanitycheck: support symbolic links in ZEPHYR_BASE

marc-hb opened the pull request:
Fix issue where sanitycheck wrongly assumed tests inside ZEPHYR_BASE
to be outside ZEPHYR_BASE and dropped the prefix in their name. This
happened when:
- ZEPHYR_BASE contains symbolic link(s), and
- relative --testcase-root argument(s) are passed

To generate unique names, TestCase.get_unique(testcase_root) first
checks whether "testcase_root" starts with ZEPHYR_BASE. Either may or
may not include symbolic links so both must be canonicalized before
comparison. While fixing this method, replace explicit forward slash
"/" and string replace with os.path.relpath() and make a couple other
simplifications and minor pydoc fixes.

Add new canonical_zephyr_base = os.path.realpath(ZEPHYR_BASE) constant
and corresponding comments and guidelines.

The most visible effect of this mismatch was sanitycheck dropping the
--testcase-root prefix from the unique name of tests inside
ZEPHYR_BASE. This means some test names could be not unique anymore
and silently overwrite each other's results, example:

 bash# cd zephyr_dir_with_symlink; export ZEPHYR_BASE=$(pwd)
 ./scripts/sanitycheck -T samples/portability/cmsis_rtos_v1 \
                       -T samples/portability/cmsis_rtos_v2

The more systematic and practical consequence (and how I actually
found this) was test outputs landing in unexpected locations.

mbolivar commented on scripts/sanitycheck:
```diff
@@ -1687,11 +1687,11 @@ class TestCase:
     def get_unique(self, testcase_root, workdir, name):
 
-        if ZEPHYR_BASE in testcase_root:
+        real_zephyr_base = os.path.realpath(ZEPHYR_BASE)
```
Why not normalize the path when ZEPHYR_BASE is assigned, once and for all?

marc-hb replied:
Good question. While not larger in lines of code, this would have massively larger side effects which may or may not be desired and/or buggy.
ZEPHYR_BASE is referred to ~ 20 times in this script but no other seems to require normalization, this small fix is enough to make everything work for me.

marc-hb replied:
There's also a consistency issue. sanitycheck is essentially a wrapper on top of the main build system and in number of places it uses ZEPHYR_BASE to generate path arguments to other tools of the build. Let's pretend my `/home/march-db` is a symbolic link to `/nfs/serverA/home/marc-hb`. I would really not like sanitycheck to invoke the build with `/nfs/serverA/home/marc-hb` while I invoke it with `/home/march-db`, that would be at least a source of confusion and deception and possibly worse.

mbolivar replied:
Hm, I don't think I understand the change well enough to review then. I'll drop my -1 and see what @nashif says :)

mbolivar reviewed (dismissed):
It is good to allow symbolic links, but I think the variable should be treated uniformly.

mbolivar reviewed (dismissed):
Marc [4:07 PM]
the bug is: I break the match when using the default value
can you "request changes" in github?
(I can't!)

marc-hb committed:
sanitycheck: support symbolic links in ZEPHYR_BASE

Fix issue where sanitycheck wrongly assumed tests inside ZEPHYR_BASE
to be outside ZEPHYR_BASE and dropped the prefix in their name. This
happened when:
- ZEPHYR_BASE contains symbolic link(s), and
- relative --testcase-root argument(s) are passed

To generate unique names, TestCase.get_unique(testcase_root) first
checks whether "testcase_root" starts with ZEPHYR_BASE. Either may or
may not include symbolic links so both must be canonicalized before
comparison. While fixing this method, replace explicit forward slash
"/" and string replace with os.path.relpath() and make a couple other
simplifications and minor pydoc fixes.

Add new canonical_zephyr_base = os.path.realpath(ZEPHYR_BASE) constant
and corresponding comments and guidelines.

The most visible effect of this mismatch was sanitycheck dropping the
--testcase-root prefix from the unique name of tests inside
ZEPHYR_BASE. This means some test names could be not unique anymore
and silently overwrite each other's results, example:

 bash# cd zephyr_dir_with_symlink; export ZEPHYR_BASE=$(pwd)
 ./scripts/sanitycheck -T samples/portability/cmsis_rtos_v1 \
                       -T samples/portability/cmsis_rtos_v2

The more systematic and practical consequence (and how I actually
found this) was test outputs landing in unexpected locations.

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

marc-hb commented:
`testcase_root` can have have symlinks too when it's the default (thx @mbolivar), pushed new version that canonicalizes both `testcase_root` and `ZEPHYR_BASE`

mbolivar commented:
> `testcase_root` can have have symlinks too when it's the default (thx @mbolivar), pushed new version that canonicalizes both `testcase_root` and `ZEPHYR_BASE`

Is `--testcase-root` the only one? What about `--board-root`? Are there any others?

marc-hb commented:
I reviewed all direct and indirect uses of `ZEPHYR_BASE` and I haven't found any that involves a path comparison or requires canonicalization for any other reason. I've tested sanitycheck (compiling hundreds of tests) with symbolic links using a fair number of its features and I didn't find any other similar issue. In case I missed one similar issue in some other feature I didn't use, the new `canonical_zephyr_base` constant will make it easier to fix. 

Keep in mind the only thing this PR fixes is test outputs landing in the wrong place (and in case of extremely bad luck overwriting each other). So it's possible that people have *already* been using sanitycheck with symbolic links for a long time without noticing any problem.

mbolivar reviewed (approved):
Seems sane to me. If there are any other issues where ZEPHYR_BASE is being used and canonical_zephyr_base should be used instead, those can be treated as separate bugs IMO.
